### PR TITLE
allow using specific jack server instance

### PR DESCRIPTION
**Pull request by djfun**
Add a new command line parameter "--jackservername <servername>", which will
be used to connect to a jack server instance with the specified name.
If the parameter is missing, connect to the default jack server.

See #1394 and #1397 

So far, I have only tested building for linux.

### CODE REVIEW

File: android/sound.cpp [hunk 1]
```diff
@@ -33,6 +33,7 @@ CSound::CSound ( void           (*fpNewProcessCallback) ( CVector<short>& psData
                  void*          arg,
                  const QString& strMIDISetup,
                  const bool     ,
+                 const QString&,
```
Comment (hoffie) [inline android/sound.cpp:36]: It's suboptimal that we have to add implementation-specific flags such as this jack option to all implementations (Windows, Android, Mac), but I can't think of any sane alternatives. Also, it follows the existing logic for jack-specific options (`bNoAutoJackConnect`), so I guess it's ok.

File: src/main.cpp [hunk 2]
```diff
@@ -532,6 +533,22 @@ int main ( int argc, char** argv )
         }
 
 
+        // Jack server name ----------------------------------------------------
```
Comment (hoffie) [inline src/main.cpp:536]: Could we only add this logic if we have jack support at all (`#ifdef WITH_SOUND`, it seems. #1408)?

(I know that the existing code for jack options does not lead by good example in this regard, we should try to be better nevertheless :))

File: src/main.cpp [hunk 3]
```diff
@@ -875,6 +893,7 @@ QString UsageArguments ( char **argv )
         "      --mutemyown       mute me in my personal mix (headless only)\n"
         "  -c, --connect         connect to given server address on startup\n"
         "  -j, --nojackconnect   disable auto Jack connections\n"
+        "      --jackservername  use specific Jack server instance\n"
```
Comment (hoffie) [inline src/main.cpp:896]: Should also be guarded by `#ifdef WITH_SOUND`, IMO.

**Review by hoffie — COMMENTED** [review]
Thanks for the PR! Looks good in general. I have added some small comments.

File: src/main.cpp (hunk 2 above)
Comment (djfun) [reply src/main.cpp:536]: I updated my commit

**Review by hoffie — APPROVED** [review]
Thanks!
Could you also add yourself to the contributors' list in https://github.com/jamulussoftware/jamulus/blob/ad12ba0543e21792b6760d6b0e1df860b3e61741/src/util.cpp#L489?

**Comment by hoffie** [comment]
Thinking some more about it, would this do the same as the `JACK_DEFAULT_SERVER` environment variable? If so, does the latter work for Jamulus? If so, I'm not sure if we should have an additional command line flag.

**Comment by djfun** [comment]
> Thinking some more about it, would this do the same as the `JACK_DEFAULT_SERVER` environment variable? If so, does the latter work for Jamulus? If so, I'm not sure if we should have an additional command line flag.

:-/ yes it works, at least in my test setup. It should probably just be mentioned somewhere for people to find it.

**Comment by pljones** [comment]
Ideally... there'd be a UI to accompany this.  Somewhere in the Device selector, in Settings:
image
(Or list the available Jack instances in the "Device" box?  That sort of fits the meaning, I guess.)

**Review by softins — DISMISSED** [review]
I've built this on a RPi, and it appears to start a jack server with the appropriate name as given, or to connect to the running `jackdbus` if the name is not specified or is given as `default`.

**Comment by softins** [comment]
> I've built this on a RPi, and it appears to start a jack server with the appropriate name as given, or to connect to the running `jackdbus` if the name is not specified or is given as `default`.

Having reviewed the discussion more closely, I have just confirmed that both these commands give exactly the same effect:
```
./Jamulus --jackservername fred
```
```
JACK_DEFAULT_SERVER=fred ./Jamulus
```
So probably better to document this somewhere rather than add a command option.

File: linux/sound.cpp [hunk 4]
```diff
@@ -28,12 +28,16 @@
...
-                        const char* jackClientName )
+                        const char* jackClientName,
+                        const char* jackServerName )
 {
     jack_status_t JackStatus;
 
     // try to become a client of the JACK server
-    pJackClient = jack_client_open ( jackClientName, JackNullOption, &JackStatus );
+    if (strlen(jackServerName) > 0)
+        pJackClient = jack_client_open ( jackClientName, JackServerName, &JackStatus, jackServerName );
```
[the remark this answers is not in the document]
Comment (pljones) [reply linux/sound.cpp:38]: And it needs `{` and `} adding to make the statement a block in the standard code style.

[the remark this answers is not in the document]
Comment (djfun) [reply linux/sound.cpp:38]: yes, sorry, I am too used to the linux kernel coding style.

**Comment by pljones** [comment]
> Yes, I was about to write the same. I don't think we should add features (especially those which need UI logic und space) without documented need by users? :)

If a feature is needed (i.e. we need to choose a Jack server instance), then it should have a UI by default, not require a command line parameter.

**Comment by softins** [comment]
Closing this as it's not needed after all, since the environment variable provides equivalent functionality for those who need it.

**Pull request closed without merging**